Multiple time integrators in an input

lindsayad opened the pull request:
Closes #19228

lindsayad committed:
Add test

lindsayad committed:
Plumbing for multiple time integrators

Closes #19228

lindsayad committed:
Don't multiply time derivatives by 2 for Crank-Nicolson

lindsayad committed:
Revert "Don't multiply time derivatives by 2 for Crank-Nicolson"

This reverts commit 572f7f6d312dcb171e280099d3d837802f42a007.

lindsayad committed:
sub-vector time derivative calculation

lindsayad commented:
relies on https://github.com/libMesh/libmesh/pull/3856 and https://github.com/libMesh/libmesh/pull/3857

lindsayad commented:
don't close

lindsayad committed:
Update to new restore_subvector API

lindsayad committed:
Add extra space per Brandon's suggestion

lindsayad committed:
Adapt modules to new time integrator APIs

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2489082/), step Docs: sync website on a50e022 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27687/site)

This comment will be updated on new commits.

lindsayad committed:
Make subvectors restartable so recover works

lindsayad committed:
Denote that DirectCentralDifference overridesSolve

lindsayad committed:
Make overridesSolve pure virtual

This forces downstream developers to remember to implement it

lindsayad committed:
Fix time integrator retrieval in Simulation.C

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2510857/), step Generate coverage on 136a365 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#27687 136a36</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>85.06%</td>
      <td>+0.01%</td>
      <td>94.21%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106298</td>
      <td>106481</td>
      <td>+183</td>
      <td>293</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18691</td>
      <td>18703</td>
      <td>+12</td>
      <td>18</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage/framework)


## Modules coverage

### Fsi

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#27687 136a36</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.23%</td>
      <td>87.23%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>328</td>
      <td>328</td>
      <td>-</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>48</td>
      <td>48</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage_diff/fsi)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage/fsi)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#27687 136a36</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.48%</td>
      <td>84.48%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17002</td>
      <td>17002</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3123</td>
      <td>3123</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage/navier_stokes)

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#27687 136a36</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.91%</td>
      <td>84.91%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>28158</td>
      <td>28159</td>
      <td>+1</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>5006</td>
      <td>5006</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage/solid_mechanics)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#27687 136a36</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.35%</td>
      <td>88.35%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13910</td>
      <td>13912</td>
      <td>+2</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1834</td>
      <td>1834</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27687/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27687/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad committed:
Address unity build failure

lindsayad committed:
Add TimeIntegrators index page

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2514190/) on 8bb4b77 : invalidated by @lindsayad

lindsayad committed:
Make changes backwards compatible with Griffin

loganharbour commented on framework/include/systems/SystemBase.h:
```diff
@@ -950,6 +943,28 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+  /**
+   * @returns All the time integrators owned by this system
+   */
+  const std::vector<std::shared_ptr<TimeIntegrator>> & getTimeIntegrators();
```
Does it make sense here to not grant write access? i.e., reconstruct and give a `std::vector<std::shared_ptr<const TimeIntegrator>>`

lindsayad replied:
Both `Transient` and `FEProblemBase` call non-const methods on the time integrators returned by this API. The vector should be constant, but the pointers shouldn't be

loganharbour commented on framework/include/timeintegrators/AStableDirk4.h:
```diff
@@ -62,6 +62,7 @@ class AStableDirk4 : public TimeIntegrator
                                 ADReal & ad_u_dotdot) const override;
   virtual void solve() override;
   virtual void postResidual(NumericVector<Number> & residual) override;
+  virtual bool overridesSolve() const override { return true; }
```
Another non-required change, but I think we do this more now via a private param in the base class

lindsayad replied:
A private data member or private method? I want the pure virtual to ensure that developers must override it. I initially missed a time integrator in the solid mechanics module which led to a bunch of diffs when I didn't have the pure virtual

loganharbour replied:
I didn't realize it was a pure virtual. I like that more. Sounds good to me.

loganharbour commented on framework/include/timeintegrators/TimeIntegrator.h:
```diff
@@ -164,6 +179,12 @@ class TimeIntegrator : public MooseObject, public Restartable
+   * Copy from one vector into another. If the time integrator has been restricted to a subset of
+   * variables, then this will selectively copy their dofs
+   */
+  void copyVector(NumericVector<Number> & from, NumericVector<Number> & to);
```
```suggestion
  void copyVector(const NumericVector<Number> & from, NumericVector<Number> & to);
```

loganharbour commented on framework/src/restart/DataIO.C:
```diff
@@ -326,7 +326,11 @@ dataStore(std::ostream & stream,
           void * context)
 {
-  mooseAssert(v, "Null vector");
+  bool have_vector = v.get();
```
Can you add a comment here as to why you needed this? As it's a bit inconsistent with other nullptr things

loganharbour commented on framework/src/timeintegrators/AStableDirk4.C:
```diff
@@ -92,7 +92,9 @@ AStableDirk4::computeTimeDerivatives()
   computeTimeDerivativeHelper(u_dot, _solution_old);
   u_dot.close();
-  _du_dot_du = 1. / _dt;
+  for (const auto i : index_range(_du_dot_du))
```
Should there be a helper for this? Looks like it's used quite a lot

loganharbour commented on framework/src/timeintegrators/ImplicitEuler.C:
```diff
@@ -32,11 +32,26 @@ ImplicitEuler::computeTimeDerivatives()
+  }
+  else
+  {
+    auto u_dot_sub = u_dot.get_subvector(_local_indices);
```
Same for this with a helper

loganharbour commented on framework/src/timeintegrators/TimeIntegrator.C:
```diff
@@ -33,26 +35,81 @@ TimeIntegrator::TimeIntegrator(const InputParameters & parameters)
 TimeIntegrator::solve()
 {
-  _nl.system().solve();
+  mooseError("Calling this method is no longer supported");
```
```suggestion
  mooseError("Calling TimeIntegrator::solve() is no longer supported");
```

loganharbour reviewed (changes requested):
Looks great. Most of this is just suggestions

lindsayad commented on framework/include/systems/SystemBase.h:
```diff
@@ -244,9 +244,9 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+  virtual std::vector<Number> & duDotDus() { return _du_dot_du; }
   virtual Number & duDotDotDu() { return _du_dotdot_du; }
-  virtual const Number & duDotDu() const { return _du_dot_du; }
+  virtual const Number & duDotDu(unsigned int var_num = 0) const;
```
It is the exact same scenario. I am happy to make the change a breaking one in order to not have these defaulted arguments

loganharbour replied:
Up to you.

lindsayad replied:
I'll keep it the way it is for now, but I've opened https://github.com/idaholab/moose/issues/28966 to keep track of this and take care of it quickly after propagating the MOOSE changes into Griffin

lindsayad committed:
Apply suggestions from code review

- Remove commented method
- `getPossiblyNullTimeIntegrator` -> `queryTimeIntegrator`
- `const` addition to `from` vector in `copyVector`
- Add method name to `mooseError` in `TimeIntegrator::solve`
- Add comment regarding handling of null vectors
- Create computeDuDotDu and duDotDuCoeff methods

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2517196/), step Clang format on b3dcdcb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27687/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27687/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f5a01db01841e690890a2a245f55ea13fefb199f`

lindsayad commented:
thanks for the review!
